### DISCUSSION TITLE
How to load data when navigating to a page

### DISCUSSION THREAD

**Discussion by chanan** [rrousselGit/riverpod#2156] [Q&A] [answered]
Hello,

I have been using async provider to save data to fire store. So, so far all my interactions have been due to tap events. Now I want to display the data so I first load it and then build the widget tree. Which requires one line to execute my controller like so:

```
ref.read(goalControllerProvider.notifier).getAllGoals(account: account);
```

When I tired that I got the error: `Tried to modify a provider while the widget tree was building`. 

Coming from a react background, I thought maybe I should wrap it in a `useEffect()` but got the same error.

What is the correct way to load data based on navigation instead of a button press? Should I load the data in the previous screen before navigation but if thats the case will deep linking work (which I am not currently using to be fair, but want to make sure I am not boxing myself in)

**Comment by chanan** [comment]
For now I moved the getAllGoals to my bottom navigation bar which works. I guess I could wrap the view with another widget that has nothing in it other than the load and that should work too. Is that the correct way to deal with this or is there a better way?

**Comment by mboyamike** [comment]
You would probably want to use a FutureProvider for such a thing.

```
final goalsProvider = FutureProvider.family<List<Goal>, Account>((ref, account) => ref.watch(goalControllerProvider).getAllGoals(account: account));
```

**Reply by mboyamike** [reply]
Then on the widget you can just watch that provider

**Comment by avalanche-tm** [comment, marked as answer]
@chanan you have to also wrap it into WidgetsBinding.instance.addPostFrameCallback to postpone change after first build. You might want to create custom hook like this to help you with boilerplate:

```dart
void useOnInit(Function action) {
  useEffect(() {
    WidgetsBinding.instance.addPostFrameCallback(
      (_) => action(),
    );
    return null;
  }, []);
}
```

and then you can call it in your Widget's build method like this:

```dart
useOnInit(() => ref.read(goalControllerProvider.notifier).getAllGoals(account: account));
```

**Reply by chanan** [reply]
Thanks so much that is exactly what I was looking for!
